proposals/p2240.md

```diff
@@ -20,30 +20,39 @@ SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 -   [Details](#details)
 -   [Rationale](#rationale)
 -   [Alternatives considered](#alternatives-considered)
-    -   [Decouple variadics from arrays](#decouple-variadics-from-arrays)
-    -   [Other array type syntaxes](#other-array-type-syntaxes)
+    -   [Support variadic variable declaration statements](#support-variadic-variable-declaration-statements)
+    -   [Support array expansion arguments](#support-array-expansion-arguments)
+    -   [Omit variadic bindings](#omit-variadic-bindings)
+        -   [Disallow pack-type bindings](#disallow-pack-type-bindings)
     -   [Fold expressions](#fold-expressions)
     -   [Allow multiple `...,` patterns in a tuple pattern](#allow-multiple--patterns-in-a-tuple-pattern)
     -   [Allow nested pack expansions](#allow-nested-pack-expansions)
-    -   [Disallow named packs](#disallow-named-packs)
-    -   [Use postfix instead of prefix operators](#use-postfix-instead-of-prefix-operators)
+    -   [Use postfix instead of prefix `...`](#use-postfix-instead-of-prefix-)
+    -   [Require parentheses around `each`](#require-parentheses-around-each)
     -   [Keyword syntax](#keyword-syntax)
-        -   [Rationale](#rationale-1)
-        -   [Discussion](#discussion)
 
 <!-- tocstop -->
 
 ## Abstract
 
 Proposes a set of core features for declaring and implementing generic variadic
-functions. The central concept is a "pack", which is a value that consists of a
-sequence of other values. Almost all operations that can apply to single values
-are generalized to apply to packs as well, by applying the operation to each
-element of the pack. Packs are created from tuples or statically-sized arrays
-using the `[:]` operator, and can be converted back to tuples with the `...,`
-operator. Variadic blocks, delimited with `...{ }`, enable iterative computation
-on packs. `[:]` and `...,` can also be used in patterns, which enables functions
-to take an arbitrary number of arguments.
+functions. The central concept is a _pack expansion_, which is a kind of
+compile-time loop over sequences called _expansion arguments_. Expansion
+arguments can be formed from tuples with the `expand` operator, or from a
+_variadic binding_, which is marked with the `each` keyword at the point of
+declaration and the point of use. Pack expansions come in several forms, with
+slightly different iteration semantics:
+
+-   `...,` iteratively evaluates an expression, and treats the values as
+    successive elements of an enclosing tuple.
+-   `...and` and `...or` iteratively evaluates a boolean expression, combining
+    the values using `and` and `or`, and ending the loop early if the underlying
+    operator short-circuits.
+-   `...{}` iteratively executes a block of statements.
+
+`...,` can also be used in patterns, where it iteratively matches the elements
+of a scrutinee tuple. In conjunction with variadic bindings, this enables
+functions to take an arbitrary number of arguments.
```

Comment:
This is great. Perhaps copy it as an intro into `variadics.md`?

Reply:
I worry that as an introduction it would put off a lot of readers, because it's so compressed and has no examples. Am I thinking about that the wrong way?

Reply:
I think it is useful as a summary of what you are going to say, giving a high level view. My expectation is that this summary will give the reader a fuzzy global view where they don't really understand everything, but they can then refine by reading further. Right now `variadics.md` is very hard to understand since it goes into depth into syntax and terminology before talking at all about semantics. So the reader has no framework for understanding the syntax when it is being explained.

I think an even better structure is:
- This summary
- The simplest example, `SumInts`, end to end. This should include the explanation of each new syntax being introduced, one at a time, and the meaning of the result given an example.
- Another example that adds one feature (`Min`), to demonstrate that feature, along with an explanation of that feature.
- Repeat until all features have been explained.

That is a big change, though, and I think it would be reasonable to just move the variadics.md document wholesale into the details section of the proposal and delete the variadics.md file from this PR. Once the proposal is approved, a follow-up PR can iterate on creating variadics.md.

Reply:
> I think it is useful as a summary of what you are going to say, giving a high level view. My expectation is that this summary will give the reader a fuzzy global view where they don't really understand everything, but they can then refine by reading further. Right now `variadics.md` is very hard to understand since it goes into depth into syntax and terminology before talking at all about semantics. So the reader has no framework for understanding the syntax when it is being explained.

I see, that makes a lot of sense. Done.

> I think an even better structure is:
> 
> * This summary
> * The simplest example, `SumInts`, end to end. This should include the explanation of each new syntax being introduced, one at a time, and the meaning of the result given an example.
> * Another example that adds one feature (`Min`), to demonstrate that feature, along with an explanation of that feature.
> * Repeat until all features have been explained.
> 
> That is a big change, though, and I think it would be reasonable to just move the variadics.md document wholesale into the details section of the proposal and delete the variadics.md file from this PR. Once the proposal is approved, a follow-up PR can iterate on creating variadics.md.

Thanks, I really like that approach. I might try to do that in this PR, because the proposal can't move forward if people don't understand it.

Reply:
Just following up here with a repeat of what I think came up in the open discussion today --

I think actually for the proposal, we've got enough context to understand it and let it move forward with the simple suggestion from Josh and for the larger change to be a follow-up to make the design docs work more broadly. We've been discussing this enough that at least for me, it's gelling nicely, and I'm inclined to not lose that momentum for a more complex change to the document. =]